### DISCUSSION TITLE
扩展 build.js，支持更多功能

### DISCUSSION THREAD

**Discussion by zahoorwang** [vue-mini/vue-mini#117] [Ideas]
现实开发中，可能会遇到如下问题：

 - [x] 不同的运行环境，灵活的环境变量；
 - [x] CSS的预处理；
 - [x] 单一文件方式写； 
 - [x] JS/TS 可能会混乱编写（虽然不建议，但要优化原来老JS项目到TS的时候可能会产生）；
 - [ ] 编译脚本的 hook，比如在全部页面级别的 wxml 中添加统一的组件，比如：`<t-toast id="toast" />` （tdesign 的）；

 其它等等的要待解决的...

为了解决上面的问题，分享一个扩展过的 [build.js](https://gist.github.com/zahoorwang/49c21ca3445deb950a7be920b4e6de58)

脚本地址：https://gist.github.com/zahoorwang/49c21ca3445deb950a7be920b4e6de58

------
### 1. 不同的运行环境，灵活的环境变量（已完成）：
使用 `dotenv` 和 `dotenv-expand` 完成此功能，借鉴了 `vite` 的 `loadEnv` 编写解决，区分了 `NODE_ENV` 的变量，只是让这个变量只干 运行模式 的区分。

需要注意以下环境变量：
- `process.env.NODE_ENV`，只有两个值：`production` 和 `development`
- `process.env.MODE`，可以自定义，如：`development`，`testing`，`production`，等等，对应的环境变量文件为：
    - `.env.development` 和 `.env.development.local`
    - `.env.testing` 和 `.env.testing.local`
    - `.env.production` 和 `.env.production.local`
    - `.env` 和 `.env.local`
    这里与 `vite` 的处理方式完全一样

也有 `prefix` 的功能，目前支持的为：`APP_`。

在 `package.json` 的 `scripts` 中可以这样写：
```json
{
  "build": "cross-env NODE_ENV=production MODE=production node build.js",
  "build:test": "cross-env NODE_ENV=production MODE=testing node build.js",
  "start": "cross-env NODE_ENV=development MODE=development node build.js"
}
```

### 2. CSS的预处理（已完成）：
纯CSS编写，稍微有点痛苦，懂得都懂...，而微信开发者工具中虽然支持了 `less` 和 `sass`，但好像是一个太监版本的（反正个人在用的时候，嵌套功能貌似没有了），坑到想哭...

这里扩展的脚本，支持了 `less` 和 `sass`，在编译过程中，支持处理了这两个预处理，不用预先安装模块，在发现没有的时候会在控制台中输入警告提示安装，当然可以混用（强烈不建议）。

暂时只支持了最基本的使用，像 `vite` 一样，可以添加变量等功能，需等待... 或者大家一起贡献...

### 3. 单一文件（已完成）：
现在编写方式，一个页面或组件都需要建立4个文件，文件有点爆炸了；大项目的话看起来相当的挺凌乱的...

这里借助了 `Vue` 的 `@vue/compiler-sfc` 来解析一个 xxx.vue （非标准 Vue 文件，只是借助格式）来处理：
- `template` 会生成为 `文件名.wxml`
- `script` 会生成为 `文件名.js`，`lang` 属性支持 `js` 和 `ts`；不支持多个，也不支持 `setup`
- `style` 会生成为 `文件名.wxss`，`lang` 属性支持 `css`，`less` 和 `scss`，其它属性不支持；同时支持多个 `style` 节点，会按顺序生成到 `wxss` 文件中
- `declare` 会生成 `文件名.json`，预置了 `lang="json"` 和 `mode="page | component"`，属性方面还未支持，考虑为以后写一个类似 `unplugin-vue-router`  缩减版功能所留

还请注意：单一文件存在的时候，会覆盖单独写的对应文件

举个🌰：
```vue
<!-- src/pages/test/index.vue  -->

<template>
  <view class="test">
    <view wx:if="{{ true }}">
      <block>{{ greet }}</block>
    </view>
  </view>
</template>

<script lang="ts">
import { definePage, ref } from '@vue-mini/core';
definePage({
  setup() {
    const greet = ref<string>('hello !');
    return { greet };
  }
});
</script>

<style lang="scss">
.test {
  position: relative;
}
</style>

<style>
.test {
  height: 100vh;
}
</style>

<declare lang="json" mode="page">
{
  "usingComponents": {}
}
</declare>

<!--
会生为：
dist/pages/test/index.json
dist/pages/test/index.js
dist/pages/test/index.wxml
dist/pages/test/index.wxss
-->
```

### 4. JS/TS 可能会混乱编写（已完成）：
这个只是把大佬写的两个脚本糅合了一下

### 5. 编译脚本的 hook：
该功能考虑中，暂时没有想到一个比较好的方案，可能需要将 `build.js` 抽象成类似 `unbuild` 或 `vite` 这样提供独立 `cli` 之后更容易实现吧～～

------
以上都在真实项目中使用了，暂时还没有发现啥问题；也希望大家一起完善，有什么问题一起讨论哈～

**Comment by yangmingshan** [comment]
太强了👍

**Comment by DeMeng33** [comment]
太强了，先学习一下怎么编译sass。
